File: code_samples/section12/example_4_interpolation_search/interpolation_search.java
package code_samples.section12.example_4_interpolation_search;

import java.io.File;                 // Represents file and directory pathnames
import java.io.FileNotFoundException; // Thrown when a file cannot be found
import java.util.*;                   // Provides collections and Scanner

/**
 * Demonstrates Interpolation Search on ordered integer data,
 * including a version that counts comparison steps.
 */
public class interpolation_search {

    // ======================================================
    // Interpolation Search (index-only)
    // ======================================================

    /**
     * Performs interpolation search on a sorted integer array.
     *
     * Interpolation search estimates the likely position of the target
     * based on its value relative to the values at the current bounds.
     * This approach works best when the data is uniformly distributed.
     *
     * @param arr     Sorted array to search
     * @param target  Value to locate
     *
     * @return Index of the target if found; otherwise -1
     */
    int interpolationSearch(int[] arr, int target) {
        int n = arr.length;        // Number of elements in the array
        if (n == 0) return -1;     // Empty array => not found

        int lo = 0, hi = n - 1;    // Search bounds (inclusive)

        // Continue searching while:
        //  - bounds are valid, and
        //  - target lies within the value range [arr[lo], arr[hi]]
        while (lo <= hi && target >= arr[lo] && target <= arr[hi]) {

            // --------------------------------------------------
            // Special case: all values in [lo, hi] are identical
            // --------------------------------------------------
            if (arr[hi] == arr[lo]) {
                // Only possible match is at index lo
                return (arr[lo] == target) ? lo : -1;
            }

            // --------------------------------------------------
            // Estimate the likely position using interpolation:
            //
            // pos = lo + (hi - lo) * (target - arr[lo]) / (arr[hi] - arr[lo])
            //
            // Casting to double avoids integer truncation during division.
            // --------------------------------------------------
            int pos = lo + (int)((double)(hi - lo) * (target - arr[lo])
                    / (arr[hi] - arr[lo]));

            // Guard against out-of-range estimates
            if (pos < lo || pos > hi) return -1;

            // Compare against the estimated position
            if (arr[pos] == target) return pos;
            else if (arr[pos] < target) lo = pos + 1; // Search right subrange
            else hi = pos - 1;                         // Search left subrange
        }

        // Target not found
        return -1;
    }

    // ======================================================
    // Interpolation Search WITH step counting
    // ======================================================

    /**
     * Performs interpolation search while counting selected comparisons.
     *
     * The step counter includes:
     *  - Checking if arr[hi] == arr[lo]
     *  - Checking if arr[lo] == target (uniform-range case)
     *  - Checking if arr[pos] == target
     *  - Checking if arr[pos] < target (to choose direction)
     *
     * Loop boundary checks are not counted—only the explicitly
     * labeled comparisons are included.
     *
     * @param arr       Sorted array to search
     * @param target    Value to locate
     * @param stepsOut  Single-element array used to return step count
     *
     * @return Index of the target if found; otherwise -1
     */
    int interpolationSearchSteps(int[] arr, int target, int[] stepsOut) {
        int steps = 0;        // Counts number of comparisons
        int n = arr.length;   // Number of elements

        // Handle empty array case
        if (n == 0) {
            stepsOut[0] = 0;
            return -1;
        }

        int lo = 0, hi = n - 1; // Search bounds

        // Continue searching while target remains within value bounds
        while (lo <= hi && target >= arr[lo] && target <= arr[hi]) {

            steps++; // Compare arr[hi] == arr[lo]
            if (arr[hi] == arr[lo]) {
                steps++; // Compare arr[lo] == target
                stepsOut[0] = steps;
                return (arr[lo] == target) ? lo : -1;
            }

            // Estimate likely position
            int pos = lo + (int)((double)(hi - lo) * (target - arr[lo])
                    / (arr[hi] - arr[lo]));

            // Guard against invalid estimates
            if (pos < lo || pos > hi) {
                stepsOut[0] = steps;
                return -1;
            }

            steps++; // Compare arr[pos] == target
            if (arr[pos] == target) {
                stepsOut[0] = steps;
                return pos;
            }

            steps++; // Compare arr[pos] < target
            if (arr[pos] < target) lo = pos + 1;
            else hi = pos - 1;
        }

        // Target not found
        stepsOut[0] = steps;
        return -1;
    }

    // ======================================================
    // Load ordered.txt relative to project directory
    // ======================================================

    /**
     * Loads the ordered integer dataset from disk.
     *
     * Expected project structure:
     *   <root>/code_samples/section12/data/ordered.txt
     *
     * @return Array of integers read from ordered.txt,
     *         or an empty array on failure
     */
    static int[] loadOrderedData() {
        // Resolve project root directory
        String root = System.getProperty("user.dir");

        // Build full path to ordered.txt
        String path = root
                + File.separator + "code_samples"
                + File.separator + "section12"
                + File.separator + "data"
                + File.separator + "ordered.txt";

        System.out.println("Attempting to read: " + path);

        List<Integer> list = new ArrayList<>();

        // Use try-with-resources to ensure Scanner is closed automatically
        try (Scanner sc = new Scanner(new File(path))) {
            while (sc.hasNextInt()) {
                list.add(sc.nextInt());
            }
        } catch (FileNotFoundException e) {
            System.out.println("Error reading ordered.txt: " + e.getMessage());
            return new int[0];
        }

        // Convert List<Integer> → primitive int[]
        int[] arr = new int[list.size()];
        for (int i = 0; i < arr.length; i++) arr[i] = list.get(i);

        return arr;
    }

    // ======================================================
    // MAIN Test Harness
    // ======================================================

    /**
     * Program entry point.
     *
     * Loads ordered data from disk and runs interpolation search tests,
     * reporting step counts for different search scenarios.
     */
    public static void main(String[] args) {

        // Load ordered dataset
        int[] arr = loadOrderedData();
        if (arr.length == 0) {
            System.out.println("Missing input file - aborting.");
            return;
        }

        interpolation_search is = new interpolation_search(); // Search instance
        int[] steps = new int[1];                             // Step counter output

        System.out.println("\n=== Interpolation Search Tests (sorted data only) ===");
        System.out.println("Loaded " + arr.length + " integers\n");

        // Test 1: Search for the first element
        int idx = is.interpolationSearchSteps(arr, arr[0], steps);
        System.out.println("Search first  (" + arr[0] + "): index=" + idx + ", steps=" + steps[0]);

        // Test 2: Search for the middle element
        int midVal = arr[arr.length / 2];
        idx = is.interpolationSearchSteps(arr, midVal, steps);
        System.out.println("Search middle (" + midVal + "): index=" + idx + ", steps=" + steps[0]);

        // Test 3: Search for the last element
        int lastVal = arr[arr.length - 1];
        idx = is.interpolationSearchSteps(arr, lastVal, steps);
        System.out.println("Search last   (" + lastVal + "): index=" + idx + ", steps=" + steps[0]);

        // Test 4: Search for a value guaranteed not to exist
        idx = is.interpolationSearchSteps(arr, 999_999, steps);
        System.out.println("Search missing (999999): index=" + idx + ", steps=" + steps[0]);
    }
}

File: code_samples/section12/example_4_interpolation_search/interpolation_search.cpp
#include <iostream>  // Provides standard input/output streams (cout)
#include <vector>    // Provides the std::vector container
#include <fstream>   // Provides file stream classes (ifstream)

using namespace std; // Allows use of standard library names without std:: prefix

// ============================================================
// Interpolation Search (index-only)
// Requires: arr sorted in ascending order
// ============================================================

/**
 * Performs interpolation search on a sorted vector of integers.
 *
 * Interpolation search is similar to binary search, but instead of always
 * selecting the midpoint, it estimates the likely index of the target based
 * on the target's value relative to the values at the current bounds.
 *
 * This approach tends to work best when values are fairly uniformly distributed.
 *
 * @param arr     Sorted vector to search
 * @param target  Value to locate
 *
 * @return Index of the target if found; otherwise -1
 */
int interpolationSearch(const vector<int>& arr, int target) {
    int n = (int)arr.size(); // Number of elements
    if (n == 0) return -1;   // Empty dataset => not found

    int lo = 0;              // Lower bound of the search range (inclusive)
    int hi = n - 1;          // Upper bound of the search range (inclusive)

    // Continue searching while:
    //  - bounds are valid, and
    //  - target is within the value range [arr[lo], arr[hi]]
    while (lo <= hi && target >= arr[lo] && target <= arr[hi]) {

        // Special case: if arr[lo] == arr[hi], all values in [lo, hi] are equal.
        // If that shared value equals target, return lo; otherwise, target can't exist.
        if (arr[hi] == arr[lo]) {
            // All values in [lo, hi] are the same
            if (arr[lo] == target) return lo;
            return -1;
        }

        // --------------------------------------------------------
        // Estimate probable position using linear interpolation:
        //
        // pos = lo + (hi - lo) * (target - arr[lo]) / (arr[hi] - arr[lo])
        //
        // Casting to double avoids integer truncation during division,
        // then the result is converted back to int for indexing.
        // --------------------------------------------------------
        int pos = lo + (int)((double)(hi - lo) * (target - arr[lo]) /
                             (arr[hi] - arr[lo]));

        // Guard: pos should generally fall within [lo, hi], but rounding or
        // unusual distributions can produce out-of-range estimates.
        if (pos < lo || pos > hi) {
            return -1;
        }

        // Compare against the estimated position
        if (arr[pos] == target) {
            return pos;            // Target found
        } else if (arr[pos] < target) {
            lo = pos + 1;          // Target must be in the right subrange
        } else {
            hi = pos - 1;          // Target must be in the left subrange
        }
    }

    // Target not found (range exhausted or target out of bounds)
    return -1;
}

// ============================================================
// Interpolation Search WITH step counting
// steps = number of comparisons
// ============================================================

/**
 * Performs interpolation search while counting selected comparisons.
 *
 * Step counting here focuses on key comparisons in the algorithm:
 *  - Checking if arr[hi] == arr[lo] (uniform-range check)
 *  - Checking if arr[lo] == target when uniform-range
 *  - Checking if arr[pos] == target
 *  - Checking if arr[pos] < target (to decide which side to continue searching)
 *
 * Note: The while-loop boundary checks (lo<=hi and target range checks)
 * are not counted here; only the labeled comparisons are counted.
 *
 * @param arr    Sorted vector to search
 * @param target Value to locate
 * @param steps  Output parameter used to record comparison count
 *
 * @return Index of the target if found; otherwise -1
 */
int interpolationSearchSteps(const vector<int>& arr, int target, int& steps) {
    steps = 0;               // Reset step counter for this search
    int n = (int)arr.size(); // Number of elements
    if (n == 0) return -1;   // Empty dataset => not found

    int lo = 0;              // Lower bound of search range
    int hi = n - 1;          // Upper bound of search range

    // Continue searching while target remains within the current value bounds
    while (lo <= hi && target >= arr[lo] && target <= arr[hi]) {

        steps++;  // compare arr[hi] == arr[lo]
        if (arr[hi] == arr[lo]) {
            // All values in range are equal; only possible match is that value
            steps++;  // compare arr[lo] == target
            if (arr[lo] == target) {
                return lo;   // Target found
            } else {
                return -1;   // Target not present
            }
        }

        // Estimate the likely position of the target
        int pos = lo + (int)((double)(hi - lo) * (target - arr[lo]) /
                             (arr[hi] - arr[lo]));

        // Guard: reject out-of-range estimated positions
        if (pos < lo || pos > hi) {
            return -1;
        }

        steps++;  // compare arr[pos] == target
        if (arr[pos] == target) {
            return pos;      // Target found
        }

        steps++;  // compare arr[pos] < target
        if (arr[pos] < target) {
            lo = pos + 1;    // Search right subrange
        } else {
            hi = pos - 1;    // Search left subrange
        }
    }

    // Target not found
    return -1;
}

// ============================================================
// Load integers from file into vector<int>
// ============================================================

/**
 * Loads integers from a text file into a vector<int>.
 *
 * @param path Path to the input file
 *
 * @return Vector of integers read from the file,
 *         or an empty vector if the file cannot be opened
 */
vector<int> loadFile(const string& path) {
    ifstream in(path);  // Open file for reading
    vector<int> arr;    // Container for loaded integers

    // Validate file open
    if (!in) {
        cout << "Error reading: " << path << "\n";
        return arr;     // Return empty vector on failure
    }

    int x; // Temporary input variable

    // Read integers until EOF or invalid input
    while (in >> x) {
        arr.push_back(x);
    }

    return arr; // Returned by value (move semantics apply)
}

// ============================================================
// Run tests: first, middle, last, missing
// ============================================================

/**
 * Executes a standardized set of interpolation search tests on ordered data.
 *
 * @param arr Sorted vector of integers to test
 */
void runTests(const vector<int>& arr) {
    cout << "=== Interpolation Search Tests (ordered data only) ===\n";
    cout << "Loaded " << arr.size() << " integers\n\n";

    int steps = 0; // Holds comparison count for each search
    int idx;       // Holds returned index

    // Test 1: Search for the first element
    idx = interpolationSearchSteps(arr, arr.front(), steps);
    cout << "Search first  (" << arr.front() << "): index=" << idx
         << ", steps=" << steps << "\n";

    // Test 2: Search for the middle element
    int midVal = arr[arr.size() / 2];
    idx = interpolationSearchSteps(arr, midVal, steps);
    cout << "Search middle (" << midVal << "): index=" << idx
         << ", steps=" << steps << "\n";

    // Test 3: Search for the last element
    int lastVal = arr.back();
    idx = interpolationSearchSteps(arr, lastVal, steps);
    cout << "Search last   (" << lastVal << "): index=" << idx
         << ", steps=" << steps << "\n";

    // Test 4: Search for a value guaranteed not to exist
    idx = interpolationSearchSteps(arr, 999999, steps);
    cout << "Search missing (999999): index=" << idx
         << ", steps=" << steps << "\n\n";
}

// ============================================================
// MAIN
// ============================================================

/**
 * Program entry point.
 *
 * Loads an ordered dataset from disk, runs interpolation search tests,
 * and exits.
 */
int main() {
    // If compiled & run from: code_samples/section12/lesson_3_interpolation_search
    // then this path points to: code_samples/section12/data/ordered.txt
    string path = "..\\data\\ordered.txt";

    // Load ordered dataset from file
    vector<int> arr = loadFile(path);

    // Validate successful file load
    if (arr.empty()) {
        cout << "Missing input file or failed to load.\n";
        return 1; // Exit with error code
    }

    // Run interpolation search benchmarks
    runTests(arr);

    return 0; // Successful program termination
}
